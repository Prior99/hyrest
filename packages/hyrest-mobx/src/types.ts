export interface Indexable<KeyType> {
    id?: KeyType;
}

export interface Target {
    value: any;
    checked?: boolean;
    type: string;
}

export interface ReactEvent {
    target: Target;
}

/**
 * A context factory, being provided to `@hasFields` to retrieve the context inside
 * the `Field` instances.
 */
export type ContextFactory<TContext> = () => TContext;
