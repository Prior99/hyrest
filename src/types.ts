export interface Answer<T> {
    readonly message?: string;
    readonly data: T;
}

export interface Params {
    [key: string]: string;
}

/**
 * This type represents all possible HTTP methods.
 */
export type HTTPMethod =
    "GET" |
    "POST" |
    "PATCH" |
    "PUT" |
    "DELETE" |
    "HEAD" |
    "OPTIONS" |
    "TRACE";

/**
 * This error will be thrown if the call to a route from the client did not succeed.
 * It transports the HTTP status code as well as the answer from the server if available.
 * Both can be `undefined` if a network failure is the reason for the error.
 */
export class ApiError extends Error {
    /**
     * The HTTP status code. Can be `undefined`.
     */
    public statusCode: number;
    /**
     * The answer from the server. Can be `undefined`.
     */
    public answer: Answer<any>;

    constructor(statusCode: number, answer: Answer<any>) {
        super()/* istanbul ignore next */;
        Error.captureStackTrace(this, ApiError);
        this.statusCode = statusCode;
        this.answer = answer;
    }
}

export type Constructable<T> = {
    new(): T;
};
