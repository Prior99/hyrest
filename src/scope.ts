import "reflect-metadata";

import { Constructable } from "./types";
import * as invariant from "invariant";
import { allKeys } from "./all-keys";
import { getTransforms } from "./transform";
import { isCustomClass } from "./validation";

export type TypeCreator = () => Constructable<any>;

export interface SpecifiedTypes {
    /**
     * Types specified for method's parameters.
     */
    params: Map<number, TypeCreator>;
    /**
     * The type specified for the property itself.
     */
    property?: TypeCreator;
}

export interface PropertyMeta {
    /**
     * The prototype of the class on which the property was decorated.
     */
    readonly target: Object;
    /**
     * The name of the property that was decorated.
     */
    readonly property: string;
    /**
     * The expected type for this property.
     */
    readonly expectedType: any | any[];
}

export class Scope {
    private included: Scope[] = [];
    private ownProperties: PropertyMeta[] = [];

    /**
     * Include all properties from another scope in this scope:
     *
     * **Example:**
     *
     * ```
     * const exampleA = createScope();
     * const exampleB = createScope().include(exampleA);
     *
     * class A {
     *     @scope(exampleA)
     *     public property1: string;
     *
     *     @scope(exampleB)
     *     public property2: string;
     * }
     *
     * console.log(exampleA.properties); // Only `property1`.
     * console.log(exampleB.properties); // Will be `property1` and `property2`.
     * ```
     * @param scopes A list of scopes to include in this scope.
     *
     * @return This instance to use this method for the builder pattern.
     */
    public include(...scopes: Scope[]): Scope {
        scopes.forEach(includedScope => this.included.push(includedScope));
        // Builder pattern: Return `this`.
        return this;
    }

    public registerProperty(property: PropertyMeta) {
        this.ownProperties.push(property);
    }

    /**
     * All properties decorated by `@scope` with this scope and all properties
     * included by using `include()`.
     */
    public get properties() {
        return this.included.reduce((result, included) => {
            result.push(...included.properties);
            return result;
        }, [ ...this.ownProperties ]);
    }

    /**
     * Will return all properties for a specific class.
     *
     * **Example:**
     *
     * ```
     * const example = createScope();
     *
     * class A {
     *     @scope(example)
     *     public property1: string;
     * }
     *
     * class B {
     *     @scope(example)
     *     public property2: string;
     * }
     *
     * console.log(example.propertiesForClass(A)); // Only `property1` is returned.
     * console.log(example.propertiesForClass(A)); // Only `property2` is returned.
     * ```
     *
     * This method takes all properties included using `include()` into account.
     *
     * @param clazz The clazz to return the properties for.
     *
     * @return All properties from the specified class.
     */
    public propertiesForClass(clazz: Function) {
        return this.properties.filter(property => {
            let current = clazz;
            while (current !== null && typeof current !== "undefined") { // tslint:disable-line
                if (property.target.constructor === current) {
                    return true;
                }
                current = Object.getPrototypeOf(current);
            }
        });
    }
}

/**
 * Create a new scope.
 *
 * @return A new scope.
 */
export function createScope() {
    return new Scope();
}

/**
 * Use this decorator to mark a property, getter or a setter on a class as belonging to a scope
 * so they are included when `dump()`ing or `populate()`ing them.
 *
 * @param scopes The list of scopes in which the decorated property should be included.
 *
 * @return A decorator which can be used to decorate getters, setters and properties.
 *
 * @see dump
 * @see populate
 */
export function scope(...scopes: Scope[]) {
    return function<T>(target: Object, property: string, descriptor?: PropertyDescriptor) {
        const expectedType = Reflect.getMetadata("design:type", target, property);
        scopes.forEach(decoratedScope => decoratedScope.registerProperty({
            target, property, expectedType,
        }));
    };
}

/**
 * Returns the specified types for a property and its parameters. Is always guaranteed to
 * return an object. If non existed, a new one is created.
 *
 * @param target The target object of which to get the specified types.
 * @param property The property on `target` on which to get the specified types.
 *
 * @return An object containing the type of the property and a map for all its parameters.
 */
export function getSpecifiedType(target: Object, property: string | symbol): SpecifiedTypes {
    let specified = Reflect.getMetadata("specifytype", target, property);
    if (typeof specified === "undefined") {
        specified = {
            params: new Map(),
        };
        Reflect.defineMetadata("specifytype", specified, target, property);
    }
    return specified;
}

/**
 * Help to specify the type of a property or paramter. Might be necessary for arrays and cyclic
 * depdendencies.
 *
 * @param factory A function with no arguments returning the desired type.
 *
 * @return A decorator for a property or a parameter.
 */
export function specify<T>(factory: TypeCreator) {
    return function(target: Object, property: string | symbol, arg3?: any) {
        const specified = getSpecifiedType(target, property);
        if (typeof arg3 === "number") {
            specified.params.set(arg3, factory);
            return;
        }
        specified.property = factory;
    };
}

export function dump<T>(dumpScope: Scope): (instance: T) => T;
export function dump<T>(dumpScope: Scope, instance: T): T;
/**
 * Dump a structure with a given scope. This can either be invoked using the curried syntax
 * (`dump(scope)(obj)` or regularly (`dump(scope, obj)`).
 * This will dump the structure as new, plain objects, including only the keys marked with the
 * `@scope` decorator and the specified scope.
 *
 * @param dumpScope The scope used for dumping.
 * @param arg2 The instance to dump.
 *
 * @return The dumped, stripped down version of the input object.
 */
export function dump<T>(dumpScope: Scope, arg2?: T): T | ((instance: T) => T) {
    function internalDump<U extends any | any[]>(instance: U): U {
        if (Array.isArray(instance)) {
            return instance.map(internalDump) as any as U;
        }
        if (typeof instance === "object" && instance !== null) {
            const propertiesForClass = dumpScope.propertiesForClass(instance.constructor);
            const keys = allKeys(instance);
            return keys.reduce((result, key) => {
                if (propertiesForClass.find(({ property }) => property === key)) {
                    (result as any)[key] = internalDump((instance as any)[key]);
                }
                return result;
            }, {}) as U;
        }
        return instance;
    }
    if (typeof arg2 !== "undefined") {
        return internalDump(arg2);
    }
    return internalDump;
}

export function populate<T>(populateScope: Scope, initialClass: Constructable<T>): (data: any) => T;
export function populate<T>(populateScope: Scope, initialClass: Constructable<T>, data: any): T;
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arrayType: Constructable<T>): (data: any) => T;
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arrayType: Constructable<T>, data: any): T;
/**
 * Populate a structure of classes using an object and a scope.
 * This will recursively create instances of all classes and populate them with the nested objects in
 * the input.
 *
 * @param populateScope The scope to populate. Only properties included in this scope will be populated.
 * @param initialClass The class to start populating with.
 *             It can also be invoked with an additional array type.
 * @param arg4 When populating an array, an additional array type must be specified as the thrid argument
 *             and the data is specified as the fourth argument instead.
 *
 * @return A curried function for use in higher order functions if `arg3` is not specified and the populated
 *         structure otherwise.
 */
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arg3?: any, arg4?: any,
): T | ((data: T) => T) {
    function internalPopulate<U extends any | any[]>(
        data: any, thisClass: any, specifiedClass: Constructable<any>,
    ): U {
        // Perform population of an array.
        if (thisClass === Array) {
            invariant(Array.isArray(data), "Structure does not match. Array expected.");
            invariant(typeof specifiedClass === "function", "Array type not specified.");
            return (data as any[]).map(element => internalPopulate(element, specifiedClass, undefined)) as any as U;
        }
        const guardedClass = thisClass || specifiedClass;
        // Ignore primitives.
        if (guardedClass === Number || guardedClass === Boolean || guardedClass === String || guardedClass === Object) {
            return data;
        }
        invariant(typeof data === "object" && !Array.isArray(data), "Structure does not match. Object expected.");
        if (typeof guardedClass !== "function") {
            throw new Error("Could not infer type. This might be due to a cyclic dependency.");
        }
        // Instanciate the classe to populate.
        const instance = new guardedClass();
        // Get a list of all properties to populate.
        const propertiesForClass = populateScope.propertiesForClass(guardedClass);
        propertiesForClass.forEach(({ property, target, expectedType }) => {
            const dataValue = (data as any)[property];
            if (typeof dataValue === "undefined") {
                return;
            }
            const specifyTypeCreator = getSpecifiedType(target, property).property;
            const specifyType = specifyTypeCreator && specifyTypeCreator();
            const transforms = getTransforms(target, property);
            const populated = internalPopulate(dataValue, expectedType, specifyType);
            if (transforms.propertyTransform) {
                (instance as any)[property] = transforms.propertyTransform(populated);
            } else {
                (instance as any)[property] = populated;
            }
        });
        return instance;
    }
    // `arg3` can either be array type or data.
    if (typeof arg3 !== "undefined") {
        // `arg3 is the array type and `arg4` is the data.
        if (typeof arg4 !== "undefined") {
            return internalPopulate(arg4, initialClass, arg3);
        }
        // If `arg3` is a `Constructable`: It's the array type and a curried version shall be returned.
        if (typeof arg3 === "function") {
            return (data: any) => internalPopulate(data, initialClass, arg3);
        }
        // `arg3 is simply the data and no array type is specified.
        return internalPopulate(arg3, initialClass, undefined);
    }
    return (data: any) => internalPopulate(data, initialClass, undefined);
}
