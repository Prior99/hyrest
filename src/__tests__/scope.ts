import { Scope, scope, createScope, dump, populate, specify, precompute } from "../scope";
import { Constructable } from "../types";

let permissive: Scope, restricted: Scope;
let A: Constructable<any>, B: Constructable<any>, C: Constructable<any>;
let a1: any, a2: any, a3: any;
let b1: any, b2: any;
let c: any;

beforeEach(() => {
    permissive = createScope();
    restricted = createScope().include(permissive);

    class AInternal {
        @scope(permissive)
        public a: string;

        @scope(restricted)
        public b: number;
    }

    class BInternal {
        @scope(permissive)
        public permissiveA: AInternal;

        @scope(restricted)
        public restrictedA: AInternal;
    }

    class CInternal {
        @scope(permissive)
        public c: string;

        @scope(restricted)
        public d: number;

        @scope(permissive) @specify(() => BInternal)
        public bs: BInternal[];
    }

    A = AInternal;
    B = BInternal;
    C = CInternal;

    a1 = new AInternal();
    a1.a = "test1";
    a1.b = 2;

    a2 = new AInternal();
    a2.a = "test3";
    a2.b = 4;

    a3 = new AInternal();
    a3.a = "test5";
    a3.b = 6;

    b1 = new BInternal();
    b1.permissiveA = a1;
    b1.restrictedA = a2;

    b2 = new BInternal();
    b2.permissiveA = a3;
    b2.restrictedA = a1;

    c = new CInternal();
    c.c = "test7";
    c.d = 42;
    c.bs = [b1, b2];
});

test("marking properties on a nested structure", () => {
    expect(restricted.properties).toMatchSnapshot();
    expect(permissive.properties).toMatchSnapshot();
});

test("marking properties on a nested structure with scopes and dumping them", () => {
    const dumpRestricted = dump(restricted, c);
    const dumpPermissive = dump(permissive, c);
    expect(dumpRestricted).toMatchSnapshot();
    expect(dumpPermissive).toMatchSnapshot();
    expect(dump(restricted)(c)).toEqual(dumpRestricted);
    expect(dump(permissive)(c)).toEqual(dumpPermissive);
});

test("populating a marked and nested structure", () => {
    const input = {
        c: "foo",
        d: 42,
        dangling: "dangling",
        bs: [
            {
                permissiveA: { a: "lorem", b: 30, dangling: "test" },
                restrictedA: { a: "ipsum", b: 40, dangling: "test" },
            },
            {
                permissiveA: { a: "dolor", b: 50, dangling: "test" },
                restrictedA: { a: "sit", b: 60, dangling: "test" },
            },
            {},
        ],
    };
    const instanceRestricted = populate(restricted, C, input);
    const instancePermissive = populate(permissive, C, input);
    expect(instanceRestricted).toMatchSnapshot();
    expect(instancePermissive).toMatchSnapshot();
    expect(instanceRestricted.constructor).toEqual(C);
    expect(instanceRestricted.bs[0].constructor).toEqual(B);
    expect(instanceRestricted.bs[0].permissiveA.constructor).toEqual(A);
    expect(populate(restricted, C)(input)).toEqual(instanceRestricted);
    expect(populate(permissive, C)(input)).toEqual(instancePermissive);
});

test("populating and dumping a circular structure", () => {
    const scope1 = createScope();
    const scope2 = createScope();
    class Circular {
        @scope(scope1)
        public property1: string;

        @scope(scope2)
        public property2: string;

        @scope(scope1)
        public circular1: Circular;

        @scope(scope2)
        public circular2: Circular;

        @scope(scope1, scope2) @specify(() => Circular)
        public circularArray: Circular[];
    }

    const input = {
        property1: "foo",
        property2: "bar",
        circular1: {
            property1: "e",
            property2: "eunt",
            circular1: { property1: "c" },
            circular2: { property2: "d" },
            circularArray: [
                {
                    property1: "a",
                    property2: "b",
                    circular1: { property2: "edipiscir" },
                    circularArray: [] as any[],
                },
            ],
        },
        circular2: {
            property1: "amet",
            property2: "sit",
            circular1: { property1: "lol" },
            circular2: { property1: "rofl", circular2: {} },
            circularArray: [] as any[],
        },
        circularArray: [
            {
                property1: "baz",
                property2: "dolor",
                circular1: {},
                circular2: {},
                circularArray: [
                    { property1: "ipsum" },
                ],
            },
        ],
    };

    const populated1 = populate(scope1, Circular, input);
    const populated2 = populate(scope2, Circular, input);
    expect(populated1).toMatchSnapshot();
    expect(populated2).toMatchSnapshot();
    const dump1_1 = dump(scope1, populated1);
    const dump2_1 = dump(scope1, populated2);
    const dump1_2 = dump(scope2, populated1);
    const dump2_2 = dump(scope2, populated2);
    expect(dump1_1).toMatchSnapshot();
    expect(dump2_1).toMatchSnapshot();
    expect(dump1_2).toMatchSnapshot();
    expect(dump2_2).toMatchSnapshot();
});

test("populating a structure with an (un-)typed array", () => {
    const scope1 = createScope();

    class Untyped {
        @scope(scope1)
        public test: string[];
    }

    class Typed {
        @scope(scope1) @specify(() => String)
        public test: string[];
    }

    const input = {
        test: ["a", "b", "c"],
    };

    expect(() => populate(scope1, Untyped, input)).toThrow();
    expect(() => populate(scope1, Typed, input)).not.toThrow();
    expect(populate(scope1, Typed, input)).toMatchSnapshot();
});

test("populating a structure with a not matching input", () => {
    const scope1 = createScope();

    class Class1 {
        @scope(scope1)
        public test: boolean;
    }

    class Class2 {
        @scope(scope1)
        public test: Class1;
    }

    expect(() => populate(scope1, Class2, { test: "a" })).toThrow();
    expect(() => populate(scope1, Class2, { test: 1 })).toThrow();
    expect(() => populate(scope1, Class2, { test: [] })).toThrow();
    expect(() => populate(scope1, Class2, { test: false })).toThrow();
    expect(() => populate(scope1, Class2, { test: { test: 3 }})).not.toThrow();
});

test("populating a structure with a not matching input", () => {
    const scope1 = createScope();

    class Class1 {
        @scope(scope1)
        public test: boolean;
    }

    class Class2 {
        @scope(scope1)
        public test: Class1;
    }

    expect(() => populate(scope1, Class2, { test: "a" })).toThrow();
    expect(() => populate(scope1, Class2, { test: 1 })).toThrow();
    expect(() => populate(scope1, Class2, { test: [] })).toThrow();
    expect(() => populate(scope1, Class2, { test: false })).toThrow();
    expect(() => populate(scope1, Class2, { test: { test: 3 }})).not.toThrow();
});

test("populating a structure with an `any` or `interface` type", () => {
    const scope1 = createScope();

    interface I {
        test: boolean;
    }

    class Class1 {
        @scope(scope1)
        public test1: any;

        @scope(scope1)
        public test2: I;
    }

    expect(populate(scope1, Class1, { test1: "a", test2: {} })).toMatchSnapshot();
    expect(populate(scope1, Class1, { test1: 1, test2: { test: false } })).toMatchSnapshot();
    expect(populate(scope1, Class1, { test1: [], test2: [] })).toMatchSnapshot();
    expect(populate(scope1, Class1, { test1: false, test2: { test: true } })).toMatchSnapshot();
    expect(populate(scope1, Class1, { test1: {}})).toMatchSnapshot();
});

test("populating a structure with a setter", () => {
    const scope1 = createScope();
    const mock = jest.fn();

    class Class1 {
        @scope(scope1)
        public set test(num: number) { mock(num); }
    }
    populate(scope1, Class1, { test: 29 });
    expect(mock).toHaveBeenCalledWith(29);
});

test("dumping a structure with a getter", () => {
    const scope1 = createScope();

    class Class1 {
        @scope(scope1)
        public get test1() { return 1; }
    }

    class Class2 extends Class1 {
        @scope(scope1)
        public get test2() { return 2; }
    }
    expect(dump(scope1, new Class2())).toEqual({
        test1: 1,
        test2: 2,
    });
});

test("dumping an array", () => {
    const scope1 = createScope();

    class Class1 {
        constructor(test: string) {
            this.test = test;
        }

        @scope(scope1)
        public test: string;
    }
    expect(dump(scope1, [
        new Class1("a"),
        new Class1("b"),
        new Class1("c"),
    ])).toEqual([
        { test: "a" },
        { test: "b" },
        { test: "c" },
    ]);
});

test("dumping null in an array", () => {
    const scope1 = createScope();

    class Class1 {}

    expect(dump(scope1, [
        null, // tslint:disable-line
    ])).toEqual([
        null, // tslint:disable-line
    ]);
});

test("populating an array", () => {
    const scope1 = createScope();

    class Class1 {
        @scope(scope1)
        public test: string;
    }
    const aInstance = new Class1();
    aInstance.test = "a";
    const bInstance = new Class1();
    bInstance.test = "b";
    const cInstance = new Class1();
    cInstance.test = "c";
    const expected = [ aInstance, bInstance, cInstance ];
    const input = [ { test: "a" }, { test: "b" }, { test: "c" } ];
    expect(populate(scope1, Array, Class1 as any, input)).toEqual(expected);
    expect((populate(scope1, Array, Class1 as any) as any)(input)).toEqual(expected);
});

test("populating a cyclic dependency", () => {
    const scope1 = createScope();

    class Class1 {
        @scope(scope1) @specify(() => Class2)
        public class2: undefined;
    }

    class Class2 {
        @scope(scope1) @specify(() => Class1)
        public class1: undefined;
    }

    expect(populate(scope1, Class1, {
        class2: {
            class1: {
                class2: {
                    class1: {},
                },
            },
        },
    })).toMatchSnapshot();
});

test("populating a cyclic dependency with no type specified", () => {
    const scope1 = createScope();

    class Class1 {
        @scope(scope1)
        public class2: undefined;
    }

    class Class2 {
        @scope(scope1)
        public class1: undefined;
    }

    expect(() => populate(scope1, Class1, {
        class2: {
            class1: {
                class2: {
                    class1: {},
                },
            },
        },
    })).toThrowErrorMatchingSnapshot();
});

test("populating and dumping a structure with a date", () => {
    const scope1 = createScope();

    class Class1 {
        constructor(date?: Date) {
            this.date = date;
        }

        @scope(scope1)
        public date: Date;
    }

    const originalDate = new Date("2017-12-16T08:35:29.390Z");
    const original = new Class1(originalDate);
    expect(dump(scope1, original)).toMatchSnapshot();
    expect(populate(scope1, Class1, { date: originalDate })).toMatchSnapshot();
    expect(populate(scope1, Class1, dump(scope1, original))).toEqual(original);
});

test("populating and dumping structure with a date from a string", () => {
    const scope1 = createScope();

    class Class1 {
        @scope(scope1) @specify(() => Date)
        public date: Date;
    }

    const dateString = "2017-12-16T08:35:29.390Z";
    const date = new Date(dateString);
    const result = populate(scope1, Class1, { date: dateString });
    expect(result).toMatchSnapshot();
    expect(typeof result.date).toBe("object");
    expect(result.date.constructor).toBe(Date);
    expect(populate(scope1, Class1, dump(scope1, result))).toEqual(result);
    expect(dump(scope1, result)).toMatchSnapshot();
});

test("dumping a structure with an array of dates", () => {
    const scope1 = createScope();

    class Class1 {
        @scope(scope1) @specify(() => Date)
        public date: Date;
    }

    class Class2 {
        @scope(scope1) @specify(() => Class1)
        public class1s: Class1[];
    }

    const date1 = new Date("2017-12-16T08:00:00Z");
    const date2 = new Date("2017-12-16T08:00:00Z");
    const instance1 = new Class1();
    instance1.date = date1;
    const instance2 = new Class1();
    instance2.date = date2;
    const class2 = new Class2();
    class2.class1s = [instance1, instance2];
    expect(dump(scope1, class2)).toMatchSnapshot();
    expect(populate(scope1, Class2, dump(scope1, class2))).toEqual(class2);
});

test("populating and dumping a @precompute getter", () => {
    const scope1 = createScope();

    class Class1 {
        private value: string;

        constructor(value?: string) {
            this.value = value;
        }

        @scope(scope1) @precompute
        public get property() {
            return this.value;
        }
    }

    const instance = new Class1("some test");
    const dumped = dump(scope1, instance);
    expect(dumped).toMatchSnapshot();
    const populated = populate(scope1, Class1, dumped);
    expect(populated.constructor).toBe(Class1);
    expect(populated).toMatchSnapshot();
});
