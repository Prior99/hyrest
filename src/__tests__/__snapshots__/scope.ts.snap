// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dumping a structure with an array of dates 1`] = `
Object {
  "class1s": Array [
    Object {
      "date": 2017-12-16T08:00:00.000Z,
    },
    Object {
      "date": 2017-12-16T08:00:00.000Z,
    },
  ],
}
`;

exports[`marking properties on a nested structure 1`] = `
Array [
  Object {
    "expectedType": [Function],
    "property": "b",
    "target": AInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "restrictedA",
    "target": BInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "d",
    "target": CInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "a",
    "target": AInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "permissiveA",
    "target": BInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "c",
    "target": CInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "bs",
    "target": CInternal {},
  },
]
`;

exports[`marking properties on a nested structure 2`] = `
Array [
  Object {
    "expectedType": [Function],
    "property": "a",
    "target": AInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "permissiveA",
    "target": BInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "c",
    "target": CInternal {},
  },
  Object {
    "expectedType": [Function],
    "property": "bs",
    "target": CInternal {},
  },
]
`;

exports[`marking properties on a nested structure with scopes and dumping them 1`] = `
Object {
  "bs": Array [
    Object {
      "permissiveA": Object {
        "a": "test1",
        "b": 2,
      },
      "restrictedA": Object {
        "a": "test3",
        "b": 4,
      },
    },
    Object {
      "permissiveA": Object {
        "a": "test5",
        "b": 6,
      },
      "restrictedA": Object {
        "a": "test1",
        "b": 2,
      },
    },
  ],
  "c": "test7",
  "d": 42,
}
`;

exports[`marking properties on a nested structure with scopes and dumping them 2`] = `
Object {
  "bs": Array [
    Object {
      "permissiveA": Object {
        "a": "test1",
      },
    },
    Object {
      "permissiveA": Object {
        "a": "test5",
      },
    },
  ],
  "c": "test7",
}
`;

exports[`populating a cyclic dependency 1`] = `
Class1 {
  "class2": Class2 {
    "class1": Class1 {
      "class2": Class2 {
        "class1": Class1 {},
      },
    },
  },
}
`;

exports[`populating a cyclic dependency with no type specified 1`] = `"Could not infer type. This might be due to a cyclic dependency."`;

exports[`populating a marked and nested structure 1`] = `
CInternal {
  "bs": Array [
    BInternal {
      "permissiveA": AInternal {
        "a": "lorem",
        "b": 30,
      },
      "restrictedA": AInternal {
        "a": "ipsum",
        "b": 40,
      },
    },
    BInternal {
      "permissiveA": AInternal {
        "a": "dolor",
        "b": 50,
      },
      "restrictedA": AInternal {
        "a": "sit",
        "b": 60,
      },
    },
    BInternal {},
  ],
  "c": "foo",
  "d": 42,
}
`;

exports[`populating a marked and nested structure 2`] = `
CInternal {
  "bs": Array [
    BInternal {
      "permissiveA": AInternal {
        "a": "lorem",
      },
    },
    BInternal {
      "permissiveA": AInternal {
        "a": "dolor",
      },
    },
    BInternal {},
  ],
  "c": "foo",
}
`;

exports[`populating a structure with an (un-)typed array 1`] = `
Typed {
  "test": Array [
    "a",
    "b",
    "c",
  ],
}
`;

exports[`populating a structure with an \`any\` or \`interface\` type 1`] = `
Class1 {
  "test1": "a",
  "test2": Object {},
}
`;

exports[`populating a structure with an \`any\` or \`interface\` type 2`] = `
Class1 {
  "test1": 1,
  "test2": Object {
    "test": false,
  },
}
`;

exports[`populating a structure with an \`any\` or \`interface\` type 3`] = `
Class1 {
  "test1": Array [],
  "test2": Array [],
}
`;

exports[`populating a structure with an \`any\` or \`interface\` type 4`] = `
Class1 {
  "test1": false,
  "test2": Object {
    "test": true,
  },
}
`;

exports[`populating a structure with an \`any\` or \`interface\` type 5`] = `
Class1 {
  "test1": Object {},
}
`;

exports[`populating and dumping a @precompute getter 1`] = `
Object {
  "property": "some test",
}
`;

exports[`populating and dumping a circular structure 1`] = `
Circular {
  "circular1": Circular {
    "circular1": Circular {
      "property1": "c",
    },
    "circularArray": Array [
      Circular {
        "circular1": Circular {},
        "circularArray": Array [],
        "property1": "a",
      },
    ],
    "property1": "e",
  },
  "circularArray": Array [
    Circular {
      "circular1": Circular {},
      "circularArray": Array [
        Circular {
          "property1": "ipsum",
        },
      ],
      "property1": "baz",
    },
  ],
  "property1": "foo",
}
`;

exports[`populating and dumping a circular structure 2`] = `
Circular {
  "circular2": Circular {
    "circular2": Circular {
      "circular2": Circular {},
    },
    "circularArray": Array [],
    "property2": "sit",
  },
  "circularArray": Array [
    Circular {
      "circular2": Circular {},
      "circularArray": Array [
        Circular {},
      ],
      "property2": "dolor",
    },
  ],
  "property2": "bar",
}
`;

exports[`populating and dumping a circular structure 3`] = `
Object {
  "circular1": Object {
    "circular1": Object {
      "property1": "c",
    },
    "circularArray": Array [
      Object {
        "circular1": Object {},
        "circularArray": Array [],
        "property1": "a",
      },
    ],
    "property1": "e",
  },
  "circularArray": Array [
    Object {
      "circular1": Object {},
      "circularArray": Array [
        Object {
          "property1": "ipsum",
        },
      ],
      "property1": "baz",
    },
  ],
  "property1": "foo",
}
`;

exports[`populating and dumping a circular structure 4`] = `
Object {
  "circularArray": Array [
    Object {
      "circularArray": Array [
        Object {},
      ],
    },
  ],
}
`;

exports[`populating and dumping a circular structure 5`] = `
Object {
  "circularArray": Array [
    Object {
      "circularArray": Array [
        Object {},
      ],
    },
  ],
}
`;

exports[`populating and dumping a circular structure 6`] = `
Object {
  "circular2": Object {
    "circular2": Object {
      "circular2": Object {},
    },
    "circularArray": Array [],
    "property2": "sit",
  },
  "circularArray": Array [
    Object {
      "circular2": Object {},
      "circularArray": Array [
        Object {},
      ],
      "property2": "dolor",
    },
  ],
  "property2": "bar",
}
`;

exports[`populating and dumping a structure with a date 1`] = `
Object {
  "date": 2017-12-16T08:35:29.390Z,
}
`;

exports[`populating and dumping a structure with a date 2`] = `
Class1 {
  "date": 2017-12-16T08:35:29.390Z,
}
`;

exports[`populating and dumping structure with a date from a string 1`] = `
Class1 {
  "date": 2017-12-16T08:35:29.390Z,
}
`;

exports[`populating and dumping structure with a date from a string 2`] = `
Object {
  "date": 2017-12-16T08:35:29.390Z,
}
`;
